pkg/solidity-utils/test/foundry/fuzz/StableMathInvariant.medusa.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import { IBasePool } from "@balancer-labs/v3-interfaces/contracts/vault/IBasePool.sol";
+import { Rounding } from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+import { InputHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/InputHelpers.sol";
+
+import { BaseMedusaTest } from "@balancer-labs/v3-vault/test/foundry/utils/BaseMedusaTest.sol";
+import { Rounding } from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import { ScalingHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/ScalingHelpers.sol";
+
+import { StableMathMock } from "../../../contracts/test/StableMathMock.sol";
+
+contract StableMathInvariantMedusaTest is BaseMedusaTest {
+    uint256 constant MIN_TOKENS = 2;
+    uint256 constant MAX_TOKENS = 8;
+    uint256 constant DEFAULT_AMP_FACTOR = 200;
+
+    StableMathMock public stableMath;
+
+    int256 internal currentInvariant;
+    int256 internal invariantWithDelta;
+
+    constructor() BaseMedusaTest() {
+        stableMath = new StableMathMock();
+    }
+
+    function optimize_currentInvariant() public view returns (int256) {
+        return currentInvariant;
+    }
+
+    function optimize_InvariantWithDelta() public view returns (int256) {
+        return -invariantWithDelta;
+    }
+
+    function property_currentInvariant() public returns (bool) {
+        return invariantWithDelta >= currentInvariant;
+    }
```

Comment:
Let's add a few comments here. I understand this makes current invariant higher and new invariant lower.

I don't think we actually need to optimize the current invariant. We just want to start from some point and not make it decrease. Also if we always do new balances, isn't this just a foundry stateless fuzz test really?

Reply:
Yes, agreed! I believe the invariant test is partially implemented in the StableMath.t.sol (solidity-utils package). Maybe, if there's some case missing in there, it can be added.

But, checking the invariant is not enough, because remove liquidity lowers the invariant. We should check if there isn't a set of operations that combined can decrease the rate of the token. It means, a stateful test that stores the first rate of the pool in the constructor, then optimizes the negative of the current rate and checks the current rate at the end, to make sure the current rate is not lower than the first rate. AddAndRemoveLiquidityStable almost does it all, but does not have the swaps in there. I believe putting the Swaps test with the Add/Remove tests should be enough.

Reply:
I think the difference is that the fuzz test just uses random data, but the medusa test will try to increase `currentInvariant` and `invariantWithDelta` decrease. This is the case that we are trying to find, that there is some currentInvariant to which if we add delta, then the invariant will be less than with delta. Am I missing something?